cylc/flow/scripts/lint.py

```diff
@@ -0,0 +1,431 @@
+"""
+from colorama import Fore
+from optparse import Values
+from pathlib import Path
+import re
+from typing import Generator
+
+from cylc.flow import LOG
+from cylc.flow.option_parsers import (
+    CylcOptionParser as COP,
+)
+from cylc.flow.cfgspec.workflow import upg, SPEC
+from cylc.flow.parsec.config import ParsecConfig
+from cylc.flow.terminal import cli_function
+
+STYLE_GUIDE = (
+    'https://cylc.github.io/cylc-doc/latest/html/workflow-design-guide/'
+    'style-guide.html#'
+)
+URL_STUB = "https://cylc.github.io/cylc-doc/latest/html/7-to-8/"
+SECTION1 = r'\[\s*{}\s*\]'
+SECTION2 = r'\[\[\s*{}\s*\]\]'
+SECTION3 = r'\[\[\[\s*{}\s*\]\]\]'
+FILEGLOBS = ['*.rc', '*.cylc']
+JINJA2_SHEBANG = '#!jinja2'
+JINJA2_FOUND_WITHOUT_SHEBANG = 'jinja2 found: no shebang (#!jinja2)'
+CHECKS_DESC = {'U': '7 to 8 upgrades', 'S': 'Style'}
+STYLE_CHECKS = {
+    re.compile(r'^\t'): {
+        'short': 'Use multiple spaces, not tabs',
+        'url': STYLE_GUIDE + 'tab-characters'
+    },
+    # Not a full test, but if a non section is not indented...
+    re.compile(r'^[^\{\[|\s]'): {
+        'short': 'Item not indented.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    #            [section]
+    re.compile(r'^\s+\[.*\]'): {
+        'short': 'Too many indents for top level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    # 2 or 4 space indentation both seem reasonable:
+    re.compile(r'^(\s|\s{3}|\s{5,})\[\[.*\]\]'): {
+        'short': 'wrong number of indents for second level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    re.compile(r'^(\s{1,3}|\s{5,7}|\s{9,})\[\[\[.*\]\]\]'): {
+        'short': 'wrong number of indents for third level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    re.compile(r'\s$'): {
+        'short': 'trailing whitespace.',
+        'url': STYLE_GUIDE + 'trailing-whitespace'
+    },
+    # Consider re-adding this as an option later:
+    # re.compile(r'^.{80,}'): {
+    #     'short': 'line > 79 characters.',
+    #     'url': STYLE_GUIDE + 'line-length-and-continuation'
+    # },
+    re.compile(r'inherit\s*=\s*[a-z].*$'): {
+        'short': 'Family name contains lowercase characters.',
+        'url': STYLE_GUIDE + 'task-naming-conventions'
+    },
+}
+
+
+def list_to_config(path_):
+    """Prettify a config list"""
+    output = ''
+    for item in path_[:-1]:
+        output += f'[{item}]'
+    output += path_[-1]
+    return output
+
+
+def get_upgrader_info():
+    """Extract info about obseletions and deprecations from Parsec Objects."""
+    conf = ParsecConfig(SPEC, upg)
+    upgrades = conf.upgrader(conf.dense, '').upgrades
+    deprecations = {}
+
+    for _, upgrades_for_version in upgrades.items():
+        for upgrade in upgrades_for_version:
+            # Set a flag indicating that a variable has been moved.
+            if upgrade['new'] is None:
+                short = (
+                    f'{list_to_config(upgrade["old"])} is not '
```

Comment:
Better visual separation IMHO
```suggestion
                    f'{list_to_config(upgrade["old"])} - not '
```